### PR TITLE
React: Add support for react18's new root API

### LINKED ISSUE #10543: Render React componentes with Concurrent Mode
**Is your feature request related to a problem? Please describe.**
I would like to test some componentes to see if is there any problem or any benefits of using Concurrent Mode.

**Describe the solution you'd like**
A way to enable the Concurrent Mode rendering in addition to the current ReactDOM.render call.

**Describe alternatives you've considered**
Not use StoryBook to try Concurrent Mode.

**Are you able to assist bring the feature to reality?**
Yes, I probably would be able to help.

**Additional context**
I think adding the option of testing the components with Concurrent Mode with storybook is a way to make more developers give it a try.

### PR DESCRIPTION
**Pull request by valentinpalkovic**
Issue: #10543 #17831

Currently, it is not possible to enable the new React Root API, which was introduced in React 17/18 (ReactDOM.createRoot). To test/use the newest concurrent features from React, it is necessary to mount a component differently. Instead of mounting it via `ReactDOM.render(reactElement, domElement)` a root has to be created (`const root = ReactDOM.createRoot`). The new API provides new methods to render/unmount components.

## What I did

The new React Root Api (ReactDOM.createRoot) allows using the
newest concurrent features of React 18. A feature flag was added to
enable/disable the new Root API via the storybook's react options.

## How to test

- [ ] Is this testable with Jest or Chromatic screenshots?
- [ ] Does this need a new example in the kitchen sink apps?
- [x] Does this need an update to the documentation?

If your answer is yes to any of these, please make sure to include it in your PR.

### CODE REVIEW

**Comment by brutus-law** [comment]
Definitely needing something like this to start upgrading to React 18 for Suspense on internal apps. Any timeline on when this is mergable?

**Comment by shilman** [comment]
Hoping to look in the next week. Thanks for your patience!

File: app/react/src/client/preview/render.tsx [hunk 1]
```diff
@@ -28,9 +37,51 @@ export const render: ArgsStoryFn<ReactFramework> = (args, context) => {
...
+      );
+    } else {
+      ReactDOM.render(node, el, () => resolve(null));
+    }
   });
 
+const unmountElement = (el: Element) => {
+  const root = nodes.get(el);
+  if (root) {
+    root.unmount();
```
Comment (shilman) [inline app/react/src/client/preview/render.tsx:61]: Do we need to remove `root` from `nodes` to avoid a memory leak?

File: app/react/src/client/preview/render.tsx [hunk 2]
```diff
@@ -28,9 +37,51 @@ export const render: ArgsStoryFn<ReactFramework> = (args, context) => {
...
+        </CallbackWrapper>
+      );
+    } else {
+      ReactDOM.render(node, el, () => resolve(null));
+    }
   });
 
+const unmountElement = (el: Element) => {
+  const root = nodes.get(el);
+  if (root) {
```
Comment (shilman) [inline app/react/src/client/preview/render.tsx:60]: can we also check for `FRAMEWORK_OPTIONS?.newReactRootApi` to make this more robust? otherwise we can possibly get burned by race conditions? 🤷‍♂️

File: docs/faq.md [hunk 3]
```diff
@@ -83,6 +83,27 @@ module.exports = {
 💡 Fast Refresh only works in development mode with React 16.10 or higher.
 </div>
 
+### How do I setup the new React Context Root API with Storybook?
+
+The new React Context Root API which was introduced in React 17/18 is an opt-in feature that can be used in Storybook React.
```
Comment (shilman) [inline docs/faq.md:88]:
```suggestion
The new [React Root API](https://reactjs.org/docs/concurrent-mode-reference.html) which was introduced in React 18 is an opt-in feature that can be used in Storybook React.
```

File: docs/faq.md [hunk 4]
```diff
@@ -83,6 +83,27 @@ module.exports = {
...
+    newReactRootApi: true,
+  },
+};
+```
+
+After enabling it, it is possible to use React's newest [concurrent features](https://reactjs.org/docs/concurrent-mode-intro.html).
+
+
+<div class="aside">
+💡 The new React Context API (React.createRoot) only works with React 17.x and above.
```
Comment (shilman) [inline docs/faq.md:104]:
```suggestion
💡 The new React Root API (React.createRoot) only works with React 18 and above.
```

File: app/react/types/index.ts [hunk 5]
```diff
@@ -7,5 +7,10 @@ export interface StorybookConfig extends BaseConfig {
   reactOptions?: {
     fastRefresh?: boolean;
     strictMode?: boolean;
+    /**
+     * Uses React 18's new root API (ReactDOM.createRoot)
+     * The new root API happens to be the gateway for accessing new features of React 18 and adds out-of-the-box improvements.
+     */
+    newReactRootApi?: boolean;
```
Comment (shilman) [inline app/react/types/index.ts:14]:
```suggestion
    newRootApi?: boolean;
```

File: app/react/src/client/preview/render.tsx [hunk 6]
```diff
@@ -28,9 +37,51 @@ export const render: ArgsStoryFn<ReactFramework> = (args, context) => {
...
+    root.unmount();
+  } else {
+    ReactDOM.unmountComponentAtNode(el);
+  }
+};
+
+const getReactRoot = (el: Element): IRoot | null => {
+  if (FRAMEWORK_OPTIONS?.newReactRootApi) {
```
Comment (shilman) [inline app/react/src/client/preview/render.tsx:68]:
```suggestion
  if (FRAMEWORK_OPTIONS?.newRootApi) {
```

File: app/react/src/client/preview/render.tsx [hunk 7]
```diff
@@ -63,6 +114,29 @@ class ErrorBoundary extends ReactComponent<{
   }
 }
 
+// Will be used to execute a callback function as soon as the React Elements are mounted.
+// This is necessary for the new React Root Api, because passing a callback function to
+// the Root API's render function is not possible anymore.
+class CallbackWrapper extends ReactComponent<{ callback: () => void }, { isDivVisible: boolean }> {
```
Comment (shilman) [inline app/react/src/client/preview/render.tsx:120]: Are there other docs which show this as a best practice, or is this something that you came up with specifically for this PR? Could you do the same thing with a `useEffect` call?

File: docs/faq.md [hunk 8]
```diff
@@ -83,6 +83,27 @@ module.exports = {
...
+### How do I setup the new React Context Root API with Storybook?
+
+The new React Context Root API which was introduced in React 17/18 is an opt-in feature that can be used in Storybook React.
+
+You can set the following properties in your `.storybook/main.js` files:
+
+```js
+module.exports = {
+  reactOptions: {
+    newReactRootApi: true,
```
Comment (shilman) [inline docs/faq.md:95]:
```suggestion
    newRootApi: true,
```

**Review by shilman — CHANGES_REQUESTED** [review]
Reviewed with @chantastic since my React18 knowledge is sorely lacking. We are big fans of this work, but have some concerns mostly about the complexity of the `CallbackWrapper` and also the potential for a memory leak with the `nodes` map. In addition some other minor nits.

Thanks so much for your patience so far and for bearing with us as we get this to a mergeable state together. Would love to hear your feedback on any of these points, since we are still learning this new feature.

File: app/react/types/index.ts (hunk 5 above)
Comment (valentinpalkovic) [reply app/react/types/index.ts:14]: done!

File: app/react/src/client/preview/render.tsx (hunk 7 above)
Comment (valentinpalkovic) [reply app/react/src/client/preview/render.tsx]: Please take a look at https://github.com/reactwg/react-18/discussions/5 (section "What about the render callback?"). 
Also, https://www.npmjs.com/package/react-callback follows almost the same idea.

Update:

I have transformed the component to be function-based

File: docs/faq.md (hunk 8 above)
Comment (valentinpalkovic) [reply docs/faq.md]: done

File: app/react/src/client/preview/render.tsx (hunk 6 above)
Comment (valentinpalkovic) [reply app/react/src/client/preview/render.tsx]: done

File: app/react/src/client/preview/render.tsx (hunk 1 above)
Comment (valentinpalkovic) [reply app/react/src/client/preview/render.tsx:64]: Good catch!

File: app/react/src/client/preview/render.tsx (hunk 7 above)
Comment (valentinpalkovic) [reply app/react/src/client/preview/render.tsx]: Update:

In my last commit I have removed the CallbackWrapper component and used another approach described here:
https://github.com/reactwg/react-18/discussions/5#discussioncomment-796012

Honestly, I don't know, whether the timing will be perfect with the `setTimeout` method. Unfortunately, we cannot use the `requestIdleCallback` method, because the browser support is not the best (https://caniuse.com/requestidlecallback).

**Review by shilman — COMMENTED** [review]
An additional question after reviewing with @kylegach @ndelangen @yannbf 😄

File: app/react/src/client/preview/render.tsx [hunk 9]
```diff
@@ -28,9 +37,50 @@ export const render: ArgsStoryFn<ReactFramework> = (args, context) => {
 
 const renderElement = async (node: ReactElement, el: Element) =>
   new Promise((resolve) => {
-    ReactDOM.render(node, el, () => resolve(null));
+    // Create Root Element conditionally for new React 18 Root Api
+    const root = getReactRoot(el);
+
+    if (root) {
+      root.render(node);
+      setTimeout(() => {
```
[the remark this answers is not in the document]
Comment (ndelangen) [reply app/react/src/client/preview/render.tsx:45]: Is `root.render(node)` async?

[the remark this answers is not in the document]
Comment (valentinpalkovic) [reply app/react/src/client/preview/render.tsx:45]: Please take a look at https://github.com/reactwg/react-18/discussions/5 (section "What about the render callback?").

And take a look at the following sub-thread. This will explain the three available methods of alternatives to the render callback in detail https://github.com/reactwg/react-18/discussions/5#discussioncomment-796012

[the remark this answers is not in the document]
Comment (valentinpalkovic) [reply app/react/src/client/preview/render.tsx:45]: A part from the sub-thread: „Use the ref callback if you want to wait until the root component commits, use requestIdleTimeout if you want to just wait for low-priority idle time, and use setTimeout if you just want to wait until after React yields the first time.“

[the remark this answers is not in the document]
Comment (valentinpalkovic) [reply app/react/src/client/preview/render.tsx:45]: > Is `root.render(node)` async?

„ In the legacy root API, you could pass render a callback that is called after the component is rendered or updated.

In the New Root API, we’ve removed this callback.

With partial hydration and progressive SSR, the timing for this callback would not match what the user expects. To avoid the confusion moving forward, we recommend using requestIdleCallback, setTimeout, or a ref callback on the root.“ - https://github.com/reactwg/react-18/discussions/5

[the remark this answers is not in the document]
Comment (valentinpalkovic) [reply app/react/src/client/preview/render.tsx:45]: @shilman Can I somehow support further? :)

[the remark this answers is not in the document]
Comment (shilman) [reply app/react/src/client/preview/render.tsx:45]: Thanks @valentinpalkovic -- I'm just waiting for the thumbs up from @chantastic who's going to do some testing of this. I definitely want to get this out as part of 6.5!

File: app/react/src/client/preview/render.tsx (hunk 7 above)
Comment (chantastic) [reply app/react/src/client/preview/render.tsx]: Hi @valentinpalkovic! Thanks for your work on this.

As mentioned earlier, I reviewed this behind-the-scenes with @shilman and raised the concern about the `CallbackWraper`. I still have reservations. So I asked in the React Working Group discussion how we could achieve this same callback timing without the complexity of the wrapper.

You can find that [sub-discussion here](https://github.com/reactwg/react-18/discussions/5#discussioncomment-2275882).

In summary, Sebastian Markbåge suggested this approach:

```js
export const CallbackWrapper: React.FC<CallbackProps> = ({
  callback,
  children
}) => {
  let once = useRef(false);
  useLayoutEffect(() = {
    if (once.current) return;
    once.current = true;
    callback();
  }, [callback]);
  return children;
}
```

Comment (valentinpalkovic) [reply app/react/src/client/preview/render.tsx]: @chantastic the CallbackWrapper was already removed a couple of iterations ago. I am not sure, whether you have reviewed the very latest version of my MR. Also, later on, I have posted some links (https://github.com/storybookjs/storybook/pull/17215#discussion_r805998060) to show up alternative solutions to the mentioned issue of not having a callback function anymore. Please take a look at my follow up comments

File: app/react/src/client/preview/render.tsx [hunk 10]
```diff
@@ -28,9 +37,50 @@ export const render: ArgsStoryFn<ReactFramework> = (args, context) => {
...
+    ReactDOM.unmountComponentAtNode(el);
+  }
+};
+
+const getReactRoot = (el: Element): IRoot | null => {
+  if (!FRAMEWORK_OPTIONS?.newRootApi) {
+    return null;
+  }
+
+  if (!(ReactDOM as any).createRoot) {
```
Comment (thure) [inline app/react/src/client/preview/render.tsx:68]: This is now throwing on `react-dom@18.0.0-rc.2` since `createRoot` is only available if it's imported as `import ReactDOM from 'react-dom/client`.

Announcement here: https://github.com/reactwg/react-18/discussions/5

File: app/react/types/index.ts [hunk 11]
```diff
@@ -7,5 +7,10 @@ export interface StorybookConfig extends BaseConfig {
   reactOptions?: {
     fastRefresh?: boolean;
     strictMode?: boolean;
+    /**
+     * Uses React 18's new root API (ReactDOM.createRoot)
+     * The new root API happens to be the gateway for accessing new features of React 18 and adds out-of-the-box improvements.
+     */
+    newRootApi?: boolean;
```
Comment (Himself65) [inline app/react/types/index.ts:14]: change the field to `suspense`?

**Comment by yannbf** [comment]
Hey @valentinpalkovic this is such a great contribution! I hope you don't mind – I'll update this branch and fix the merge conflicts, and attempt to add an E2E test for the flags you added.

File: app/react/src/client/preview/render.tsx (hunk 10 above)
Comment (yannbf) [reply app/react/src/client/preview/render.tsx:68]: Might be worth noting how Nextjs checks whether it should use React18 apis or not: https://github.com/vercel/next.js/blob/9bff48b9d5998455c9c038d7f5834a58bb9f429f/packages/next/server/config.ts#L736

**Comment by joepuzzo** [comment]
Now that React18 is out is this looking to get merged in the near future ?

File: app/react/types/index.ts (hunk 11 above)
Comment (LZL0) [reply app/react/types/index.ts:14]: I believe that would make sense.

**Review by w3bdesign — APPROVED** [review]

**Comment by shilman** [comment]
Yes, we'll try to expedite this and get it into alpha ASAP. Thanks for your patience everybody!

File: app/react/src/client/preview/render.tsx (hunk 10 above)
Comment (shilman) [reply app/react/src/client/preview/render.tsx:68]: Thanks @valentinpalkovic for your persistence and patience here. I'll also spend some time on it this weekend and hope we can get it merged early next week. 🙏

**Comment by shilman** [comment]
Note: I've merged #17853 as a temporary improvement for React18 users

**Review by shilman — COMMENTED** [review]
@valentinpalkovic if you're available, I'd love to discuss this on the Storybook discord https://discord.gg/storybook and figure out how to get it over the line together

File: app/react/src/client/preview/render.tsx [hunk 12]
```diff
@@ -28,9 +39,54 @@ export const render: ArgsStoryFn<ReactFramework> = (args, context) => {
...
+    ReactDOM.unmountComponentAtNode(el);
+  }
+};
+
+const canUseReactRoot =
+  gte(reactDomVersion, '18.0.0') || coerce(reactDomVersion)?.version === '18.0.0';
+
+const getReactRoot = (el: Element): IRoot | null => {
+  if (!FRAMEWORK_OPTIONS?.newRootApi) {
+    return null;
```
Comment (shilman) [inline app/react/src/client/preview/render.tsx:70]:
```js
import { once } from '@storybook/client-logger';

if(canUseReactRoot) {
  once.info('To use the new React18 API in Storybook, set reactOptions.newRootApi: true` in your main.js https://..../MIGRATION.md#react18-support')
}
```

Comment (shilman) [reply app/react/src/client/preview/render.tsx:70]: I'll take care of this one

Comment (brutus-law) [reply app/react/src/client/preview/render.tsx:70]: Unrelated observer so let me know if this is the improper place but..... 

Are you sure the config prop should be called `newRootApi`? I ask, because when later versions of React comes, this will be the standard api from my understanding, which means eventually this prop will need to be removed or switched to support old api as React depreciates the older way. 

From a less-obstructive-upgrade-path mentality, shouldn't it be something like:
- Check react version
- If < 18, assume the old way
- else, have the config actually be "oldRootApi" and allow using the old way if its configured, otherwise use the new way by default?

That way, there's an understanding that when you switch to 18 with storybook, you get the new api and if people need the old api until they get things upgraded they can enable it as such.  

Otherwise, I see down the line either cutting people off cold turkey when react no longer supports the old api, or having to transition people from newRootApi to oldRootApi anyway which would require another config change down the road.

Just a thought if it's worth considered.

Comment (shilman) [reply app/react/src/client/preview/render.tsx:70]: @valentinpalkovic we discussed @brutus-law 's  "opt out" approach as a possibility, but I think you had some reservations. Can you summarize them here so we can get more feedback? Thanks! 🙏

Comment (valentinpalkovic) [reply app/react/src/client/preview/render.tsx:70]: I had just some reservations of not having a flag at all. Instead of having a opt-in flag, we can rather opt-out by using a flag like „oldRootApi“. Makes totally sense!

**Comment by shilman** [comment]
@valentinpalkovic getting this error in CI. Do you know what's going on here?

```
ERR! => Failed to build the preview
ERR! /tmp/storybook/app/react/dist/esm/client/preview/render.js
ERR! Cannot find module: 'react-dom/client'. Make sure this package is installed
```

**Comment by ndelangen** [comment]
After looking at this together with @valentinpalkovic I opened this PR:
https://github.com/storybookjs/storybook/pull/17890

**Review by ndelangen — APPROVED** [review]
code looks good to me

File: app/react/src/server/framework-preset-react-dom-hack.ts [hunk 13]
```diff
@@ -0,0 +1,22 @@
+import { readJSON } from 'fs-extra';
```
Comment (Himself65) [inline app/react/src/server/framework-preset-react-dom-hack.ts:1]: I think you just need `JSON.parse(await readFile("filename.json", "utf8"))` ...

Add one dependence only for this feature is costly

**Comment by shilman** [comment]
@valentinpalkovic
- i'll update the snapshots to get the unit tests passing
- i'm talking to @IanVS about updating the vite builder to do a similar trick to what you and @ndelangen have done for webpack

File: app/react/src/client/preview/render.tsx [hunk 14]
```diff
@@ -26,10 +31,56 @@ export const render: ArgsStoryFn<ReactFramework> = (args, context) => {
...
+      setTimeout(() => {
+        resolve(null);
+      }, 0);
+    } else {
+      ReactDOM.render(node, el, () => resolve(null));
+    }
   });
+};
+
+const canUseNewReactRootApi = reactDomVersion.startsWith('18');
```
Comment (SukkaW) [inline app/react/src/client/preview/render.tsx:50]:
```suggestion
const canUseNewReactRootApi = reactDomVersion.startsWith('18') || reactDomVersion.startsWith('0.0.0');
```

Which adds support for `react@exprimental`.

File: app/react/src/server/framework-preset-react-dom-hack.ts [hunk 15]
```diff
@@ -0,0 +1,22 @@
...
+// this should be removed once we drop support for react-dom < 18
+
+export async function webpackFinal(config: Configuration) {
+  const reactDomPkg = await readJSON(require.resolve('react-dom/package.json'));
+
+  return {
+    ...config,
+    plugins: [
+      ...config.plugins,
+      reactDomPkg.version.startsWith('18')
```
Comment (SukkaW) [inline app/react/src/server/framework-preset-react-dom-hack.ts:14]:
```suggestion
      reactDomPkg.version.startsWith('18') || reactDomPkg.version.startsWith('0.0.0')
```

And here.

**Comment by ndelangen** [comment]
From what I understand of vite, it will simply never run into the problem?

**Comment by IanVS** [comment]
>From what I understand of vite, it will simply never run into the problem?

Unfortunately, it [does](https://app.circleci.com/pipelines/github/storybookjs/storybook/24910/workflows/18ca9829-8fde-4997-afb4-01e042da6707/jobs/373237).    See https://github.com/vitejs/vite/issues/6007.  I'm experimenting with detecting the user's version of react within the vite builder, and adding `react-dom/client` to the `optimizeDeps.exclude` list, to see if that helps.  But I'm not certain it will.

File: app/react/src/server/framework-preset-react-dom-hack.ts (hunk 13 above)
Comment (valentinpalkovic) [reply app/react/src/server/framework-preset-react-dom-hack.ts:1]: `fs-extra` is used in other parts of the storybook ecosystem, therefore @ndelangen decided to just use it without thinking too much about the bundle size. For sure, we can remove the fs-extra part, if JSON.parse(await readFile) doesn't introduce a different behavior. @ndelangen What do you think?

**Comment by IanVS** [comment]
I have a hacky POC working for the vite-builder.  I'll put up a PR for it later this evening, but at least it looks like it's a solvable problem.

**Review by shilman — APPROVED** [review]
@valentinpalkovic @ndelangen tremendous job on this!! 🎉

File: app/react/src/server/framework-preset-react-dom-hack.ts (hunk 13 above)
Comment (shilman) [reply app/react/src/server/framework-preset-react-dom-hack.ts:1]: Merging this for now to keep things moving. Feel free to address this in a followup PR. I agree we should be trimming down our dependencies wherever possible if it doesn't add complexity, but could go either way on this

**Comment by shilman** [comment]
Also, this breaks the vite builder, but @IanVS is on top of this and we expect to get it fixed soon

**Comment by nareshbhatia** [comment]
@valentinpalkovic, @shilman, there is a subtle issue in Valentin's repo that I can't put my finger on. It builds fine as is and both react app and storybook run fine. However if I delete package-lock.json and build from scratch then I get this error when trying to run storybook:

```
ModuleNotFoundError: Module not found: Error: Can't resolve '@mdx-js/react' in '/Users/naresh/projects/cra-with-storybook-valentinpalkovic/src/stories'
    at /Users/naresh/projects/cra-with-storybook-valentinpalkovic/node_modules/webpack/lib/Compilation.js:2016:28
    at /Users/naresh/projects/cra-with-storybook-valentinpalkovic/node_modules/webpack/lib/NormalModuleFactory.js:798:13
    at eval (eval at create (/Users/naresh/projects/cra-with-storybook-valentinpalkovic/node_modules/webpack/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:10:1)
    at /Users/naresh/projects/cra-with-storybook-valentinpalkovic/node_modules/webpack/lib/NormalModuleFactory.js:270:22
    at eval (eval at create (/Users/naresh/projects/cra-with-storybook-valentinpalkovic/node_modules/webpack/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:9:1)
    at /Users/naresh/projects/cra-with-storybook-valentinpalkovic/node_modules/webpack/lib/NormalModuleFactory.js:434:22
    at /Users/naresh/projects/cra-with-storybook-valentinpalkovic/node_modules/webpack/lib/NormalModuleFactory.js:116:11
    at /Users/naresh/projects/cra-with-storybook-valentinpalkovic/node_modules/webpack/lib/NormalModuleFactory.js:670:25
    at /Users/naresh/projects/cra-with-storybook-valentinpalkovic/node_modules/webpack/lib/NormalModuleFactory.js:855:8
    at /Users/naresh/projects/cra-with-storybook-valentinpalkovic/node_modules/webpack/lib/NormalModuleFactory.js:975:5
    at /Users/naresh/projects/cra-with-storybook-valentinpalkovic/node_modules/neo-async/async.js:6883:13
    at /Users/naresh/projects/cra-with-storybook-valentinpalkovic/node_modules/webpack/lib/NormalModuleFactory.js:958:45
    at finishWithoutResolve (/Users/naresh/projects/cra-with-storybook-valentinpalkovic/node_modules/enhanced-resolve/lib/Resolver.js:312:11)
    at /Users/naresh/projects/cra-with-storybook-valentinpalkovic/node_modules/enhanced-resolve/lib/Resolver.js:386:15
    at /Users/naresh/projects/cra-with-storybook-valentinpalkovic/node_modules/enhanced-resolve/lib/Resolver.js:435:5
    at eval (eval at create (/Users/naresh/projects/cra-with-storybook-valentinpalkovic/node_modules/enhanced-resolve/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:16:1)
```

I even tried to pin the version of storybook from "^6.5.0-alpha.64" to "6.5.0-alpha.64" but still can't get the build to run from scratch.

This is making it impossible to replicate Valentin's setup in my own repo where I performed the two steps as recommended and then substituting "^6.4.22" with "6.5.0-alpha.64". When I run storybook, I get the above error.

The other thing I tried was to replace "^6.5.0-alpha.64" with "next" so that npm picks up the new beta version (which it is doing), but still getting the same error as above.

Anything that I am missing?

**Comment by valentinpalkovic** [comment]
@nareshbhatia Could you tell me which node and npm version you are using? I try to replicate it then. And can you provide the new package-lock.json file. Then I can take a closer  comparison on both versions.

**Comment by nareshbhatia** [comment]
Sure, @valentinpalkovic.

```
$ node -v
v16.13.2
$ npm -v
8.4.0
```

Here's the package.json with the storybook versions pinned.

```
{
  "name": "storybook-test-react-18",
  "version": "0.1.0",
  "private": true,
  "dependencies": {
    "@testing-library/jest-dom": "^5.16.4",
    "@testing-library/react": "^13.1.1",
    "@testing-library/user-event": "^13.5.0",
    "@types/jest": "^27.4.1",
    "@types/node": "^16.11.27",
    "@types/react": "^18.0.6",
    "@types/react-dom": "^18.0.2",
    "react": "^18.0.0",
    "react-dom": "^18.0.0",
    "react-scripts": "5.0.1",
    "typescript": "^4.6.3",
    "web-vitals": "^2.1.4"
  },
  "scripts": {
    "start": "react-scripts start",
    "build": "react-scripts build",
    "test": "react-scripts test",
    "eject": "react-scripts eject",
    "storybook": "start-storybook -p 6006",
    "build-storybook": "build-storybook"
  },
  "eslintConfig": {
    "extends": [
      "react-app",
      "react-app/jest"
    ]
  },
  "browserslist": {
    "production": [
      ">0.2%",
      "not dead",
      "not op_mini all"
    ],
    "development": [
      "last 1 chrome version",
      "last 1 firefox version",
      "last 1 safari version"
    ]
  },
  "devDependencies": {
    "@babel/core": "^7.17.9",
    "@storybook/addon-actions": "6.5.0-alpha.64",
    "@storybook/addon-essentials": "6.5.0-alpha.64",
    "@storybook/addon-interactions": "6.5.0-alpha.64",
    "@storybook/addon-links": "6.5.0-alpha.64",
    "@storybook/builder-webpack5": "6.5.0-alpha.64",
    "@storybook/manager-webpack5": "6.5.0-alpha.64",
    "@storybook/react": "6.5.0-alpha.64",
[package-lock.txt](https://github.com/storybookjs/storybook/files/8554076/package-lock.txt)

    "@storybook/testing-library": "^0.0.11",
    "babel-loader": "^8.2.5",
    "webpack": "^5.72.0"
  }
}
```

package-lock.json is attached.

**Comment by nareshbhatia** [comment]
[package-lock.txt](https://github.com/storybookjs/storybook/files/8554088/package-lock.txt)

Sorry, attachment did not work in last message. Here it is again.

**Comment by nareshbhatia** [comment]
Any update, @valentinpalkovic? Were you able to reproduce the issue?

**Comment by IanVS** [comment]
I think this should be a new issue, instead of a conversation in a PR that a lot of people are subscribed to.  Would you mind creating one, ideally with a link to a github repo that reproduces the problem?

**Comment by nareshbhatia** [comment]
> I think this should be a new issue, instead of a conversation in a PR that a lot of people are subscribed to. Would you mind creating one, ideally with a link to a github repo that reproduces the problem?

New issue created: https://github.com/storybookjs/storybook/issues/18094

**Comment by ackvf** [comment]
How do I switch to the new createRoot api? I am still getting this error with storybok 6.5.10.

**Comment by IanVS** [comment]
@ackvf This is an old PR with a lot of people subscribed to it.  Please feel free to join our Discord (discord.gg/storybook) to ask for help, or open an issue clearly describing the problem, if you think there is a bug and don't see an existing issue tracking it.
